### ISSUE TITLE
Growing the Scene Tree the right side up

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#363]
### **Describe the project you are working on:**
Without exceptions, this applies to every single Godot project.

### **Describe the problem or limitation you are having in your project:**
Understanding an learning the Logic behind the omnipresent metaphor of the tree structure took me way too long to understand. This is for the most part because in the the actual application the tree seems more like growing roots and the rendering sorting works from the bottom up.

This makes it really hard to communicate issues, especially when it comes to render sorting and node hierarchy. It also works directly against the otherwise brilliant tree/root terminology which even finds it's way into the API. Because it is there and because it's so fundamental to how Godot works, there is unfortunately no way around but having to explain this upside down model to newcomers.

### **Describe how this feature / enhancement will help you overcome this problem or limitation:**

- Something rendered "on top" will also appear "on top" in the node hirachy,
- The root will be at a place where you would expect roots to be
- the Node tree will grow how Trees normally grow

**--> Much easier time for newcomers, much easier time to communicate between community members, much better working (otherwise great) metaphor**
### 
**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![tree_rightsideup](https://user-images.githubusercontent.com/47016402/71934519-edaa9b00-31a4-11ea-8a5a-65c8e3539250.png)

**Describe implementation detail for your proposal (in code), if possible:**
Unfortunately not possible for me.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Don't think so.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It makes little sense communication wise if only a few people have this. Addons also make little sense for newcomers, they will learn and experience the engine without addons first.

**Comment by Duroxxigar** [comment]
Going through with this would make Godot the outlier. Root is the top most level node. This is even true in math and in programming.

**Comment by Kequc** [comment]
I wouldn't mind it being one way or the other, but I do like it. I'd welcome this change if it were implemented.

Is there any reason for the existing top to bottom presentation?

**Comment by Duroxxigar** [comment]
@Kequc Because that is how it is presented in general programming and math.

**Comment by Master-J** [comment]
I don't think that would be a good idea, godot's tree structure refers to the mathematical/programming concept of trees (which has always been presented in a top to bottom fashion) and not the vegetal one (which grows from bottom to top).

You also have to consider that godot process the nodes of the scene tree recursively, so it makes sense that the ui elements being at the depest levels are rendered on top, inverting the way the tree is presented would maybe help ui design easier for newbies but on the other side make the execution flow of the tree less straigtforward.

personal note : i never saw a software letting you build hierarchies of files/objects starting from the bottom of the screen, i think it would be a terrible ux design choice (then it could just be a toggle in the editor's setting for those who want that type of ordering, but it should remain optional and disabled by default).

**Comment by golddotasksquestions** [comment]
@Master-J 
The current way makes sense if you look at the Scene Panel and think of what is displayed there like a script and the parser running down the scrpt from top to bottom. 
But this is not a script it's a graphic. It informs me as a user about the nodes rendered in the viewport (and those not rendered). In the Scene Panel they are currently displayed in the opposite order than the viewport. It's also contradicting the Tree-branch-leaf metaphor. (Unless you know a tree grows into the earth)
I accept that this contradiction has tradition. That does not make it any less absurd. 
Especially in Godot case of the Scene Panel, I think this could be reconsidered. It's seems like I am in a vast minority with that opinion here, though. Which is ok.

**Comment by Feniks-Gaming** [comment]
I agree that Tree represents math concept of a trea not actual living tree.

On top of that it would make the least important Nodes the easiest to access and the most important Nodes the hardest to access. 

In your tree your root node is the most important and the most often access node, then you have children of a node like sprites collision boxes then Nodes you hardly ever access like your timers or random containers and script holders. With tree being upside down  it would make your random Timer76 the most top node of a tree pushing your root node far far down making it huge scrolling hassle every time.

**Comment by golddotasksquestions** [comment]
There is no rule or reason to have the math model of a tree depicted in only one particular way if something else would make more sense. _Because_ it is an abstract you can display it in any shape. 
[This](https://www.youtube.com/watch?v=8SD-GAC8NYk) is a math tree just as much much as [this](https://en.wikipedia.org/wiki/Nested_set_model#/media/File:NestedSetModel.svg) or [this](http://de.gbglassware.com/Content/upload/2018244948/201806261114418333573.jpg) or [this](https://pro-heraldica.de/wp-content/uploads/2011/03/Stammbaum-Scheele1.jpg)

> On top of that it would make the least important Nodes the easiest to access and the most important Nodes the hardest to access.

Can you explain this further? 
This proposal is just about the display of the tree in the Scene Panel, noting else. As you can see in the graphic, the tree structure is exactly the same. The only difference would be, that if you have two Sprites of UI elements and you go to the scene panel to move Sprite1 on top of Sprite2, Sprite 1 would equally appear on top of Sprite2 in the viewport. Nothing more nothing less.

> With tree being upside down it would make your random Timer76 the most top node of a tree pushing your root node far far down making it huge scrolling hassle every time.

I also don't get that. The amount of scrolling you would do would be absolutely identical. Currently the root is aligned to the top, in my proposal it is aligned to the bottom of the Scene Panel. The amount of nodes hidden depends on the size of your Scene Panel box either way. There is absolutely no difference at on what nodes would be visible in the scene panel. All it would do is to display hirachical sorting in the Scene Panel as it does in the viewport by default.

**Comment by Master-J** [comment]
> With tree being upside down it would make your random Timer76 the most top node of a tree pushing your root node far far down making it huge scrolling hassle every time.

I think the point he's trying to make is that you unconsciously treat the nodes at the top to be the most important ones of your hierarchies. And that understandable because for decades every software that has to display hierarchies of things do that from to to bottom. 

I understand that the all "tree" metaphor may be misleading for total newbies in godot/programming but sometimes you just have to accept it and live with it (and if an arguably misleading term is all someone need to start being totally lost i wish him good luck because there are things even more complex to learn/figure out).

Currently, the way the nodes are displayed in the scene panel makes sense (it won't if you're think biological trees, metaphors.....we got that) and that's how it is also done in most engines like Unity (this one i know, well). Taking Unity's example, the UI elements will also be rendered from top to bottom.

I guess you won't be getting a lot of interest with your proposal, still, if you really think this must be added to the engine, you can always open a pull request to add an editor setting that swaps the order the scene panel is displaying the scene tree.

**Comment by Feniks-Gaming** [comment]
> I think the point he's trying to make is that you unconsciously treat the nodes at the top to be the most important ones of your hierarchies. And that understandable because for decades every software that has to display hierarchies of things do that from to to bottom.

That is exactly my point when you open a file structure on computer it looks like this 

![1](https://user-images.githubusercontent.com/30668026/71998978-7a526900-3238-11ea-95e7-919f7cede0f3.PNG)

You can clearly see that the most important part of this tree is my `Local Disc C` if I delete this somehow everything goes. The least important part of this tree is something like Resources or `Setup FIles` 

Anyone who ever used computer is used to this design. It's natural to almost anyone even when I want to go back I click `go up to parent folder`

![2](https://user-images.githubusercontent.com/30668026/71999304-0fedf880-3239-11ea-9031-6c11930acdcf.PNG)

Your design goes against past 40 years of any software design and user experience.

**Comment by golddotasksquestions** [comment]
Thanks everyone for explaining your points further.

**Comment by Calinou** [comment]
I'll close this proposal due to lack of support from the community. Thanks for participating :slightly_smiling_face:

**Comment by DevCurke** [comment]
Archived, but this is so interesting, I dare to comment on it anyway.

I take this quote, because it represents the most answers here in a short sentence:
> [@Kequc](https://github.com/Kequc) Because that is how it is presented in general programming and math.

Though this is UX, not math or programming. So kiss your dev bias goodbye!

Fair:
Known patterns are reasonable to use. (The first digital treelist was rendered over 60 years ago.)
Reasonable for common people b2c products.

But this is a professional software - a place where rules can be bended.
Plus it is for games. Games invent new rules all the time!

You see: it is allowed to do it.
Also, @golddotasksquestions made a well prepared and competend presented idea. I appreciate that on many levels.
Everyone just explained. Why not think about it?

The real reason why "Tree" lists are top down is the international reading direction from top to bottom.
And thats the ONLY reason. Its just a habit.

The reason for that habit is actually writing by hand.
The List though is not written, its displayed.
No mess with ink, no rewringting everything, no hand covering parts of it while writing...

We can all agree on displaying is tremendously different then writing sth.

When we talk about reading... ever read a dropdown that actually drops upwards?
That point doesn't need to be explained.

--------------

Now the interesting part:

The Treelists idea is drill down, keeping order and sometimes: inheritance rules or drag&drop.
Thats a language, not a rule. It can be done different.

Changing the direction is not a problem. But it also doesn't improove much besides to fitting it to a metaphoric name.
If you experience it giving you a clearer picture of the content, it is worth turning it around, because it improves what it is supposed to do.

But there is more to reach for.

I actually hate treelists, because I don't want to drill down, search or be overwelmed by information.
I want to directly interact with the right ones.
TL's don't help so much with that. In godot it's at least a filter, but a bad one.
Here is why and my suggestion:

What Godot actually lacks is showing all the relations of an object, not just the hierachy around it.
It also gives the wrong impression that the scene treelist IS the game or complete context of an object, but it isn't.
There can always be other references that are not in the tree but use the object or are used by it.
This is not expressible by hierachies so a tree is just the wrong matrix for the problem.

Thats a HUGE! UX flaw. Basically misinformatiorming and confusing the user. This shoots up adaption time, mental load and production time and may lead to terrible errors or destructive behaviour.

What we need is a Network. At least as addition to the tree, the object of interest in its center and all connected & relevant around it.

------

TL;DR

To stay in metaphors: 

A mycelium, not a tree.

That should have been the answer 6 years ago.
Still have a tree.
